File: src/probability/dice.py
from dataclasses import dataclass
from typing import Union, List, Dict
from .distribution import Distribution, d6, d3

@dataclass(frozen=True)
class DiceFormula:
    dice: int  # Number of dice
    sides: int  # Number of sides per die
    modifier: int  # Fixed modifier to add

    @staticmethod
    def constant(value: int) -> 'DiceFormula':
        """Create a formula for a constant value"""
        return DiceFormula(0, 1, value)

    def __str__(self) -> str:
        base = f"{self.dice if self.dice > 1 else ''}D{self.sides}"
        if self.modifier > 0:
            return f"{base}+{self.modifier}"
        elif self.modifier < 0:
            return f"{base}{self.modifier}"
        return base
    
    def roll(self) -> Distribution[int]:
        """Roll the dice and return a distribution of results"""
        if self.dice == 0:
            return Distribution.singleton(self.modifier)
        
        # Roll one die
        single_die = Distribution.uniform(list(range(1, self.sides + 1)))
        
        # Combine multiple dice
        result = single_die
        for _ in range(self.dice - 1):
            result = result.bind(lambda x: single_die.map(lambda y: x + y))
            
        # Add modifier
        if self.modifier != 0:
            result = result.map(lambda x: x + self.modifier)
            
        return result

    def is_variable(self) -> bool:
        """Return True if this formula involves dice rolls"""
        return self.dice > 0

    def to_dict(self) -> Dict[str, int]:
        """Convert to JSON-serializable dictionary"""
        return {
            "dice": self.dice,
            "sides": self.sides,
            "modifier": self.modifier
        }

    def max_possible(self) -> int:
        """Return the maximum possible value for this formula"""
        return self.dice * self.sides + self.modifier

    def min_possible(self) -> int:
        """Return the minimum possible value for this formula"""
        return self.dice + self.modifier

    @staticmethod
    def from_dict(data: Dict[str, int]) -> 'DiceFormula':
        """Create from JSON-serializable dictionary"""
        return DiceFormula(
            dice=data["dice"],
            sides=data["sides"],
            modifier=data["modifier"]
        )

# Common formulas
D6 = DiceFormula(1, 6, 0)
D3 = DiceFormula(1, 3, 0)
TWO_D6 = DiceFormula(2, 6, 0) 

File: haskell/src/Probability/Dice.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE NoFieldSelectors #-}
{-# LANGUAGE OverloadedRecordDot #-}
{-# LANGUAGE OverloadedStrings #-}

module Probability.Dice 
    ( DiceFormula(..)
    , constant
    , roll
    , isVariable
    , d6
    , d3
    , twoD6
    ) where

import GHC.Generics (Generic)
import Data.Aeson
import Probability.Distribution (Distribution)
import qualified Probability.Distribution as D

import Data.Aeson.Types (withObject, withText, (.:), (.=), object)
import Data.Text (Text)

-- | Represents a dice roll formula like "2d6+1"
data DiceFormula = DiceFormula
    { dice :: Int     -- ^ Number of dice
    , sides :: Int    -- ^ Number of sides per die
    , modifier :: Int -- ^ Fixed modifier to add
    } deriving (Eq, Generic)

instance ToJSON DiceFormula where
    toJSON DiceFormula{..} = object
        [ "dice" .= dice
        , "sides" .= sides
        , "modifier" .= modifier
        ]

instance FromJSON DiceFormula where
    parseJSON = withObject "DiceFormula" $ \v -> DiceFormula
        <$> v .: "dice"
        <*> v .: "sides"
        <*> v .: "modifier"

-- | Create a constant value (0 dice)
constant :: Int -> DiceFormula
constant value = DiceFormula 
    { dice = 0
    , sides = 1
    , modifier = value 
    }

mkDiceFormula :: Int -> Int -> Int -> DiceFormula
mkDiceFormula dice sides modifier = DiceFormula
    { dice = dice
    , sides = sides
    , modifier = modifier
    }

-- | Roll the dice and return a distribution of results
roll :: DiceFormula -> Distribution Int
roll DiceFormula{..} 
    | dice == 0 = D.singleton modifier
    | otherwise = addModifier $ combineDice dice singleDie
  where
    -- Distribution for a single die
    singleDie = D.uniform [1..sides]
    
    -- Combine multiple dice
    combineDice 1 dist = dist
    combineDice n dist = D.bind dist $ \x ->
        D.map (+ x) (combineDice (n-1) dist)
    
    -- Add the modifier if non-zero
    addModifier dist
        | modifier == 0 = dist
        | otherwise = D.map (+ modifier) dist

-- | Check if this formula involves dice rolls
isVariable :: DiceFormula -> Bool
isVariable DiceFormula{..} = dice > 0

-- | Show dice formula in standard notation (e.g. "2D6+1")
instance Show DiceFormula where
    show DiceFormula{..} = 
        let base = if dice > 1 
                  then show dice ++ "D" ++ show sides
                  else "D" ++ show sides
        in case modifier of
            0 -> base
            m | m > 0 -> base ++ "+" ++ show m
            m -> base ++ show m

-- Common formulas
d6 :: DiceFormula
d6 = DiceFormula 1 6 0

d3 :: DiceFormula
d3 = DiceFormula 1 3 0

twoD6 :: DiceFormula
twoD6 = DiceFormula 2 6 0 